pkg/scalers/nsq_scaler.go

```diff
@@ -0,0 +1,358 @@
+package scalers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+
+	"github.com/go-logr/logr"
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	"github.com/kedacore/keda/v2/pkg/scalers/scalersconfig"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+type nsqScaler struct {
+	metricType v2.MetricTargetType
+	metadata   nsqMetadata
+	httpClient *http.Client
+	logger     logr.Logger
+}
+
+type nsqMetadata struct {
+	NSQLookupdHTTPAddresses  []string `keda:"name=nsqLookupdHTTPAddresses,  order=triggerMetadata;resolvedEnv"`
+	Topic                    string   `keda:"name=topic,                    order=triggerMetadata;resolvedEnv"`
+	Channel                  string   `keda:"name=channel,                  order=triggerMetadata;resolvedEnv"`
+	DepthThreshold           int64    `keda:"name=depthThreshold,           order=triggerMetadata;resolvedEnv, default=10"`
+	ActivationDepthThreshold int64    `keda:"name=activationDepthThreshold, order=triggerMetadata;resolvedEnv, default=0"`
+
+	triggerIndex int
+}
+
+const (
+	nsqMetricType = "External"
+)
+
+func NewNSQScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting scaler metric type: %w", err)
+	}
+
+	logger := InitializeLogger(config, "nsq_scaler")
+
+	nsqMetadata, err := parseNSQMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing NSQ metadata: %w", err)
+	}
+
+	return &nsqScaler{
+		metricType: metricType,
+		metadata:   nsqMetadata,
+		httpClient: kedautil.CreateHTTPClient(config.GlobalHTTPTimeout, true),
+		logger:     logger,
+	}, nil
+}
+
+func (m nsqMetadata) Validate() error {
+	if len(m.NSQLookupdHTTPAddresses) == 0 {
+		return fmt.Errorf("no nsqLookupdHTTPAddresses given")
+	}
+
+	if m.Topic == "" {
+		return fmt.Errorf("no topic given")
+	}
+
+	if m.Channel == "" {
+		return fmt.Errorf("no channel given")
+	}
```

Comment:
Are these checks required @wozniakjan ? The entry isn't optional

Reply:
Those seemed fine to remove 👍 I tested removing `channel` from the ScaledObject in the nsq e2e test and saw logs like this in the controller:

```
2024-11-05T00:06:15Z    ERROR   scale_handler   error resolving auth params     {"type": "ScaledObject", "namespace": "nsq-test-ns", "name": "nsq-test-so", "triggerIndex": 0, "error": "error parsing NSQ metadata: error parsing nsq metadata: missing required parameter \"channel\" in [triggerMetadata resolvedEnv]"}
```